src/server/ns_turn_server.c

```diff
@@ -1294,6 +1294,12 @@ static int handle_turn_allocate(turn_turnserver *server, ts_ur_super_session *ss
           }
         }
 
+        if (server->is_draining) {
+          // Don't allow new allocations if we are draining
+          *err_code = 300; // 300 (Try Alternate): The client should contact an alternate server for this request.
```

Comment:
Try alternate error needs to carry an alternate server ipaddress or domain

https://datatracker.ietf.org/doc/html/rfc8489#section-10

Reply:
Well, I say "needs" but i guess i'm not 100% sure on that. Could be that you don't *strictly* speaking need it.

But if any alternatives are configured in the config file, probably those should be included in the message?

Reply:
Well the "USE-ALTERNATE" flag is supposed to be the mechanism for the turnserver to say "I can't service your request, but these other servers probably could".

Like you, my use of coturn involves DNS lookup, not raw ipaddresses, so i find it really annoying that the only use of ALTERNATE-DOMAIN attribute is worded as being only allowed for DTLS / TLS connections, when it realistically should work for any type of connection.

My concern with 403 forbidden is that the server might be construed as saying "You are not authorized, bad credentials" even though that's not actually the correct one.

Insufficient capacity has only the problem of the client potentially thinking that the server is unhappy about ports ( even-port / reservation-token ), therefore it may burn an additional round-trip trying to talk to the server before it gives up and moves on.

Unfortunately we can't invent our own error codes, as they would not be comprehended by clients.

Reply:
note that the alternate server configuration option already exists

https://github.com/coturn/coturn/blob/fbe07c4a16f4e07d3ad199674016868657c4b29b/examples/etc/turnserver.conf#L574-L611

so you'd only need to read from that and populate, according to the socket type.

Reply:
RFC8489 - Does use the term MUST:

The server MUST include at least
   one ALTERNATE-SERVER attribute in the error response, which MUST
   contain an IP address of the same address family as the source IP
   address of the request message.

So I could just add the configured value, but I'm wondering if we might be better off with a different error code.  Any other opinions?

Reply:
I will add the alternate server from configuration to the 300 response.

Reply:
Turns out that's not really possible.  Right now the alternate-server configuration option enables a "redirection" mode, and all Allocate requests to this coturn instance get 300 responses.  I'm thinking it will be best stay away from the 300 response then.  I'm thinking to go with the 403 response - it's description sounds appropriate and it is in a TURN RFC.  It should not be mistaken for auth errors according to the description, 401's are for that.